bundle/phases/deploy.go

```diff
@@ -124,60 +126,87 @@ is removed from the catalog, but the underlying files are not deleted:`
 	return approved, nil
 }
 
-// The deploy phase deploys artifacts and resources.
-func Deploy(outputHandler sync.OutputHandler) bundle.Mutator {
+func deployCore(ctx context.Context, b *bundle.Bundle) diag.Diagnostics {
 	// Core mutators that CRUD resources and modify deployment state. These
 	// mutators need informed consent if they are potentially destructive.
-	deployCore := bundle.Defer(
-		bundle.Seq(
-			bundle.LogString("Deploying resources..."),
-			terraform.Apply(),
-		),
-		bundle.Seq(
-			terraform.StatePush(),
-			terraform.Load(),
-			apps.InterpolateVariables(),
-			apps.UploadConfig(),
-			metadata.Compute(),
-			metadata.Upload(),
-			bundle.LogString("Deployment complete!"),
-		),
-	)
+	cmdio.LogString(ctx, "Deploying resources...")
+	diags := bundle.Apply(ctx, b, terraform.Apply())
+
+	// following original logic, continuing with sequence below even if terraform had errors
+
+	diags = diags.Extend(bundle.ApplySeq(ctx, b,
+		terraform.StatePush(),
+		terraform.Load(),
+		apps.InterpolateVariables(),
+		apps.UploadConfig(),
+		metadata.Compute(),
+		metadata.Upload(),
+	))
+
+	// TODO: we should not output this happy message if there are errors
```

Comment:
This is different from current behaviour, can we keep it as it is now?

Reply:
This keeps the current behaviour, but current behaviour is wrong - if terraform.Apply fails it still prints the message. I'm not fixing it here to keep this PR as much pure refactoring as possible.

Reply:
Not exactly the same, today it won't print deployment complete if StatePush or Load will fail for example. Even if terraform.Apply fails, it might have created some resources from the state already and `Deployment (partially) complete`. But that's not the case if we wait to push the state.

Reply:
Good point, 8c212011aced2ff79335721d2ccbd7be0071833c